File: ml-benchmark/synduce/constraints/bst/from_list_max.f
Inductive Tree = leaf Int | node {Int, Tree, Tree};
Inductive List = elt Int | cons {Int, List};

min = \a: Int. \b: Int. if < a b then a else b;
max = \a: Int. \b: Int. if < a b then b else a;

tmin = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf w -> w
  | node {w, l, r} -> min w (min (f l) (f r))
  end
);

tmax = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    leaf w -> w
  | node {w, l, r} -> max w (max (f l) (f r))
  end
);

is_bst = fix (
  \f: Tree -> Bool. \t: Tree.
  match t with
    leaf w -> true
  | node {w, l, r} -> and (and (>= w (tmax l)) (<= w (tmin r))) (and (f l) (f r))
  end
);

cat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | elt w -> cons {w, y}
  end
);

repr = fix (
  \f: Tree -> List. \t: Tree.
  match t with
    leaf x -> elt x
  | node {x, l, r} -> cat (f l) (cons {x, f r})
  end
);

spec = fix (
  \f: List -> Int. \t: List.
  match t with
    elt x -> x
  | cons {h, t} -> max h (f t)
  end
);

target = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    leaf x -> leaf x
  | node {a, l, r} -> node {a, l, f r} /*avoid the recursion of l*/
  end
);

main = \t: Tree. if is_bst t then spec (repr (target t)) else 0;

File: ml-benchmark/synduce/constraints/bst/from_list_max.ml
type tree = Leaf of int | Node of int * tree * tree
type llist = Elt of int | Cons of int * llist

let min a b = if a < b then a else b
let max a b = if a < b then b else a

let rec tmin t =
  match t with
  | Leaf w -> w
  | Node (w, l, r) -> min w (min (tmin l) (tmin r))

let rec tmax t =
  match t with
  | Leaf w -> w
  | Node (w, l, r) -> max w (max (tmax l) (tmax r))

let rec is_bst t =
  match t with
  | Leaf w -> true
  | Node (w, l, r) ->
      (w >= tmax l) && (w <= tmin r) && is_bst l && is_bst r

let rec cat x y =
  match x with
  | Cons (h, t) -> Cons (h, cat t y)
  | Elt w -> Cons (w, y)

let rec repr t =
  match t with
  | Leaf x -> Elt x
  | Node (x, l, r) -> cat (repr l) (Cons (x, repr r))

let rec spec t =
  match t with
  | Elt x -> x
  | Cons (h, t) -> max h (spec t)

val target: tree -> tree compress
let rec target t =
  match t with
  | Leaf x -> Leaf x
  | Node (a, l, r) -> Node (a, l, target r)

let program t =
  if is_bst t then spec (repr (target t)) else 0